org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/ConnectionStreamFactory.java

```diff
@@ -87,6 +145,102 @@ public OutputStream getOutputStream() throws IOException {
 
 	}
 
+	public class ConcurrentInputStream extends InputStream {
```

Comment:
I would rename all `Concurrent*` streams we create to something like `Piped*`. On Linux the nio.channels API didn't allow reading without a corresponding write so my custom implementation made sense to be named "Conurrent".